libs/libskynet/src/skylinkValidate.ts

```diff
@@ -0,0 +1,125 @@
+  }
+  const pathElems = path.split("/");
+  for (let i = 0; i < pathElems.length; i++) {
+    if (pathElems[i] === ".") {
+      return "path cannot have a . element";
+    }
+    if (pathElems[i] === "..") {
+      return "path cannot have a .. element";
+    }
+    if (pathElems[i] === "") {
+      return "path cannot have an empty element, cannot contain //";
+    }
+  }
+  return null;
+}
+
+// validateSkyfileMetadata checks whether the provided metadata is valid
+// metadata for a skyfile.
+function validateSkyfileMetadata(metadata: any): string | null {
+  // Check that the filename is valid.
+  if (!("Filename" in metadata)) {
+    return "metadata.Filename does not exist";
+  }
+  if (typeof metadata.Filename !== "string") {
+    return "metadata.Filename is not a string";
+  }
+  const errVSP = validateSkyfilePath(metadata.Filename);
+  if (errVSP !== null) {
+    return addContextToErr(errVSP, "metadata.Filename does not have a valid path");
+  }
+
+  // Check that there are no subfiles.
+  if ("Subfiles" in metadata) {
+    // TODO: Fill this out using code from
+    // skymodules.ValidateSkyfileMetadata to support subfiles.
+    return "cannot upload files that have subfiles";
+  }
+
+  // Check that the default path rules are being respected.
+  if ("DisableDefaultPath" in metadata && "DefaultPath" in metadata) {
+    return "cannot set both a DefaultPath and also DisableDefaultPath";
+  }
+  if ("DefaultPath" in metadata) {
+    // TODO: Fill this out with code from
+    // skymodules.validateDefaultPath to support subfiles and
+    // default paths.
+    return "cannot set a default path if there are no subfiles";
+  }
+
+  if ("TryFiles" in metadata) {
+    if (!metadata.TryFiles.IsArray()) {
+      return "metadata.TryFiles must be an array";
+    }
+    if (metadata.TryFiles.length === 0) {
+      return "metadata.TryFiles should not be empty";
+    }
+    if ("DefaultPath" in metadata) {
+      return "metadata.TryFiles cannot be used alongside DefaultPath";
+    }
+    if ("DisableDefaultPath" in metadata) {
+      return "metadata.TryFiles cannot be used alongside DisableDefaultPath";
+    }
+    // TODO: finish the TryFiles checking using skymodules.ValidateTryFiles
+    return "TryFiles is not supported at this time";
+  }
+  if ("ErrorPages" in metadata) {
+    // TODO: finish using skymodules.ValidateErrorPages
+    return "ErrorPages is not supported at this time";
+  }
+
+  return null;
+}
+
+// validSkylink returns true if the provided Uint8Array is a valid skylink.
+// This is an alias for 'parseSkylinkBitfield', as both perform the same
+// validation.
+function validSkylink(skylink: string | Uint8Array): boolean {
+  // If the input is a string, convert it to a Uint8Array.
+  let u8Skylink: Uint8Array;
+  if (typeof skylink === "string") {
+    const [buf, err] = b64ToBuf(skylink);
+    if (err !== null) {
+      return false;
+    }
+    u8Skylink = buf;
+  } else {
+    u8Skylink = skylink;
+  }
```

Comment:
this looks like could be a function
`parseSkylinkAsUint8(skylink: string | Uint8Array)`

Reply:
meh

Reply:
smaller functions make it easier to read and test the code and reduce complexity which in turn reduces a chance for a bug - that's not something javascript specific fwiw

it's not a required change though although it would make the code better

Reply:
I know that there is value to splitting out functions and testing them independently, but I think there's a resolution where that makes sense - here we've got 9 lines of code with a low overall complexity (one if-else with one error check), and it feels like a lot of extra work to split things out at that resolution.

In general I look at splitting out functions when complexity is high.